Convert MC Settings page to React

tpaksu opened the pull request:
Fixes #2947

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

This PR 

- Converts the Multi-currency settings page to a single React component 
- Removes the settings saved by `WC_Settings_Page` implementation
- The PR doesn't involve any (important) visual changes, it changes the base that the setting page runs on. 
- Also it contains some code refactoring by moving the related screens in their own folders, 
- adds tests for each component, 
- removes the PHP unit tests for the old settings implementation.

The new folder structure for the components is like this:

```tree
client/
├─ multi-currency/
│  ├─ multi-currency-settings/
│  │  ├─ enabled-currencies-list
│  │  ├─ store-settings
│  ├─ single-currency-settings
```

Multi currency settings component:

![image](https://user-images.githubusercontent.com/3295/134332416-309de6f2-c1c8-40c1-895d-2c22c3927f66.png)

Single currency settings component:

![image](https://user-images.githubusercontent.com/3295/134332712-c65cf065-aad2-4091-9f17-32a991ecadd3.png)


<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

##### Automatic testing

* Run `npm run test:js multi-currency``


##### Manual testing

* Navigate to `WooCommerce > Settings > Multi-Currency` to see the MC settings page.
* Add some currencies, remove some currencies
* Change store settings and save, refresh the page and check if the settings show the correct values
* The workflows should be the same with the previous version, but without page refreshes.
* Also check the MC onboarding testing scenarios on #2786 if something in this PR breaks it.

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
-->

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

jessepearson commented:
@tpaksu I haven't looked at the code yet, but I noticed something in testing. On the main settings page the _Save changes_ button refreshes the page, but on the single currency page it does not. When the refresh happens I get a _Leave site? Changes you made may not be saved._ pop up. Is it possible to have the save button save and display a snackbar instead of refreshing the page?

tpaksu commented:
@jessepearson it shouldn't refresh the page, did you rebuild the assets?

jessepearson commented:
> it shouldn't refresh the page, did you rebuild the assets?

That was the first thing I did after switching to the branch, `npm ci`. I've even refreshed the page with dev tools open so there's no cache of files.

jessepearson commented:
Okay, I figured it out, it was an error on my end. I had the branch locally, but it wasn't up to date. I was able to `git pull`, then `npm ci`, and now everything is good. :disappear:

jessepearson commented on client/multi-currency/multi-currency-settings/store-settings/index.js:
```diff
@@ -0,0 +1,162 @@
+									}
+									data-testid={ 'enable_storefront_switcher' }
+									label={ __(
+										'Add a currency switcher to the cart widget',
```
The text here appears to be the old text, it should be what's included in the Storefront integration, including the link.
https://github.com/Automattic/woocommerce-payments/blob/develop/includes/multi-currency/StorefrontIntegration.php#L58-L63

tpaksu replied:
Addressed in 69ca957.

jessepearson commented on includes/multi-currency/Settings.php:
```diff
@@ -57,410 +47,33 @@ public function __construct( MultiCurrency $multi_currency ) {
-		$exchange_rate_options = [
-			'automatic' => sprintf(
-				/* translators: %1$s: default currency rate, %2$s: new currency exchange rate, %3$s: time rates were last updated. */
-				__( 'Fetch rate automatically. Current rate: %1$s = %2$s (Last updated: %3$s)', 'woocommerce-payments' ),
```
It looks like this text is no longer used and the text from the original design is used instead, looks like it was probably missed in the previous PR. This was changed due to the rates are not updated hourly. Also the `format_last_updated_date` method was used to display the date in the site's date & time format so that the merchant wouldn't have to do any time zone conversions and would be able to tell if something was broken if it had been days since an update.

tpaksu replied:
Yes, I've taken the Figma's as the reference for the strings in the previous (Single settings page) PR, and there was the date shown as the last update indicator, The datetime comes in as a timestamp (UTC) and I used `moment` library to show it in user's timezone in that PR. So if the rate won't update in days, still the user will see that and investigate it.

tpaksu replied:
Addressed in 16ca714.

jessepearson reviewed (changes requested):
Added a couple questions, well, the same question twice.
A couple text strings should be updated, as it appears they were reverted in a previous PR.

I think one of the largest disadvantages to moving to React from the legacy settings is the ability to easily add new settings via filters. For instance, the Storefront checkbox now lives in `client/multi-currency/multi-currency-settings/store-settings/index.js`, whereas before it was added via a filter here: https://github.com/Automattic/woocommerce-payments/blob/fix/2947-convert-mc-settings-to-react/includes/multi-currency/StorefrontIntegration.php#L48-L73 . Since the setting is moved, it should be removed from `StorefrontIntegration`.

tpaksu commented on client/multi-currency/multi-currency-settings/enabled-currencies-list/index.js:
```diff
@@ -0,0 +1,133 @@
+import EnabledCurrenciesListItem from './list-item';
+import EnabledCurrenciesListItemPlaceholder from './list-item-placeholder';
+import EnabledCurrenciesModal from './modal';
+import { createInterpolateElement } from '@wordpress/element';
```
Addressed in d867e17.

tpaksu commented on client/multi-currency/multi-currency-settings/store-settings/index.js:
```diff
@@ -0,0 +1,162 @@
+
+import { useStoreSettings } from 'wcpay/data';
+import SettingsSection from 'wcpay/settings/settings-section';
+import { createInterpolateElement } from '@wordpress/element';
```
Addressed in d867e17.

tpaksu commented:
Thanks for the detailed review @jessepearson ! 

> I think one of the largest disadvantages to moving to React from the legacy settings is the ability to easily add new settings via filters. 

I couldn't agree more, but having the single currency settings in React and having the Store settings as `WC_Settings_Page` options will cause confusion too. We needed to keep the whole either as a `WC_Settings_Page`, or as a React component, which I think having it in React would provide UI consistency, shared component usage, better room for development, etc. I believe one day we'll have something like the WP hooks inside React too. 

> For instance, the Storefront checkbox now lives in client/multi-currency/multi-currency-settings/store-settings/index.js, whereas before it was added via a filter here: fix/2947-convert-mc-settings-to-react/includes/multi-currency/StorefrontIntegration.php#L48-L73 . Since the setting is moved, it should be removed from StorefrontIntegration.

I removed it and updated the tests in 7b838ab and 25aedb5. Thanks for the heads up!

jessepearson commented:
Found something else on the single currency settings page, the preview rounds down instead of up.
Store currency GBP
Enabled currency JPY
Manual rate of 151.12
Price rounding 100
Preview GBP 210
Preview JPY should be 31800, but it's 31700

Also, did this have proper decimal and thousands separators before converting it to React?

![Screen Shot 2021-09-23 at 2 56 54 PM](https://user-images.githubusercontent.com/4634416/134559293-9774c164-18bb-4c00-bd10-19095d3dee5f.png)

tpaksu commented:
Fixed the rounding down instead up issue in e172ecc.

> Also, did this have proper decimal and thousands separators before converting it to React?

I don't remember this, I'll need to check, but I think it was calculated with JS, I don't think it used the proper formatting.

jessepearson commented:
@tpaksu I looked back at the PR that initially changed the single currency settings page, and it appears a couple other things were lost.

* If the preview was NaN it displayed "Please enter a valid number" instead of NaN.
* It appears it tried to display the amount according to the proper locale.

PR is here: https://github.com/Automattic/woocommerce-payments/pull/2875/files#diff-718668a15eb0cb2e166e36d849230ad4a69c2698485907aa36bad412bf10610dL118

These don't break functionality, but they give a better UX.

tpaksu commented:
> If the preview was NaN it displayed "Please enter a valid number" instead of NaN.

Updated the method with isNaN check and that string. 

> It appears it tried to display the amount according to the proper locale.

I've changed the currency formatter to use the extended version of the same formatter `wcpay/currency/utils@formatCurrency`. Now it tries to display the amount according to the proper locale, with a little change, it uses the decimal sep, thousands sep from the store currency as it was decided to keep visual consistency on every currency on backend.

daquinons commented:
I've tested it and it's working great, well done @tpaksu 🎉

jessepearson reviewed (approved):
Thanks for all of the changes from my nit picking, @tpaksu !
This is great 😁 
:shipit:
